### ISSUE TITLE
Closing previously idle connection early

### ISSUE THREAD

**Issue by dankinder** [tylerb/graceful#83]
Hi, finding odd behavior here and think it's a bug. On Centos6, go version go1.6.2 linux/amd64. Sample program at bottom. Basically all this does is create the server with a handler that artificially takes 2s to process, with a graceful timeout of 5s.

Then it sends a synchronous request, reads the response (so no shutdown involved yet), then another in a goroutine and calls `Stop()`.

Expected behavior: with a 5s Timeout on graceful.Server, it should let the running request (which should take 2s) finish successfully.
Expected output:
`Error (after 2.005738855s): <nil>`

Actual behavior: though the program takes 2s to actually quit, the request terminated immediately.
Actual output:
`Error (after 102.433484ms): Get http://localhost:3001/: dial tcp [::1]:3001: getsockopt: connection refused`

Note that the first synchronous request makes a difference. If you comment that out, then it behavior as expected. I suspect this has something to do with `srv.Server.ConnState` not handling StateActive.

```go
package main

import (
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"sync"
	"time"

	"github.com/tylerb/graceful"
)

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
		time.Sleep(2 * time.Second)
		fmt.Fprintf(w, "done")
	})

	s := &graceful.Server{
		Server: &http.Server{
			Handler: mux,
		},
		Timeout: 5 * time.Second,
	}

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()

		lp, err := net.Listen("tcp", ":3001")
		if err != nil {
			panic(fmt.Sprintf("Failed to listen: %v", err))
		}
		s.Serve(lp)
	}()

	// Ensure we are listening on the port
	time.Sleep(100 * time.Millisecond)

	// Make a sample first request. The connection will be left idle.
	resp, err := http.Get("http://localhost:3001/")
	if err != nil {
		panic(fmt.Sprintf("first request failed: %v", err))
	}
	ioutil.ReadAll(resp.Body)
	resp.Body.Close()

	start := time.Now()
	wg.Add(1)
	go func() {
		defer wg.Done()

		_, err := http.Get("http://localhost:3001/")
		fmt.Printf("Error (after %v): %v\n", time.Now().Sub(start), err)
	}()

	// Ensure the request goes out
	time.Sleep(100 * time.Millisecond)

	s.Stop(5 * time.Second)

	wg.Wait()
}
```

**Comment by mb0** [comment]
i think it is correct. The doc for Server.Timeout says:

> Timeout is the duration to allow **outstanding** requests to survive before forcefully terminating them.

**Comment by tylerb** [comment]
@F21 any thoughts here? I'm swamped and can't dig in to this right now.

**Comment by dankinder** [comment]
@mb0 the second request being made in this example code is an outstanding request isn't it? We fire it off in a goroutine, give 100ms to ensure it is in-flight, then call `Stop()` on the server. Since the handler takes 2s to run and the timeout is 5s, I would expect the handler to finish and the program overall to end after about 2s.

**Comment by tylerb** [comment]
@dankinder try changing https://github.com/tylerb/graceful/blob/master/graceful.go#L253 to `case http.StateNew, http.StateActive:` and see if that solves your issue.

**Comment by mb0** [comment]
@dankinder well waiting some time does not ensure the request was actually made. My guess is that the second request fires after the listener is closed. connection refused sure looks like it.

it was just my first guess after a glance in the early morning. have you tried with a dedicated waitgroup to rule that out?

**Comment by dankinder** [comment]
@tylerb unfortunately adding `http.StateActive` didn't solve it. So much for my theory.

@mb0 unfortunately to be absolutely sure that the connection was established I would need to hook into the underlying RoundTripper, but I am pretty confident that isn't the issue... I just increased all the timeouts by 10x (which gives the request a full second to establish a connection) and it behaves the same.

**Comment by mb0** [comment]
@dankinder yes sorry i just checked and was wrong.

**Comment by tylerb** [comment]
I don't have time right now to try and reproduce this (and may not even be able to on Mac), but at this point I would start putting prints in various key places to understand how things are flowing, such as where connections are added and removed in the connstate hook.

**Comment by dankinder** [comment]
Working on a unit test that exacerbates this right now.

**Comment by mb0** [comment]
ok i found the solution. sorry for just dropping the diff here. i sadly have no time to setup a pull request
```

diff --git a/graceful.go b/graceful.go
index 4cb4790..150df58 100644
--- a/graceful.go
+++ b/graceful.go
@@ -246,6 +246,7 @@ func (srv *Server) Serve(listener net.Listener) error {
        // Track connection state
        add := make(chan net.Conn)
        idle := make(chan net.Conn)
+       active := make(chan net.Conn)
        remove := make(chan net.Conn)
 
        srv.Server.ConnState = func(conn net.Conn, state http.ConnState) {
@@ -254,6 +255,8 @@ func (srv *Server) Serve(listener net.Listener) error {
                        add <- conn
                case http.StateIdle:
                        idle <- conn
+               case http.StateActive:
+                       active <- conn
                case http.StateClosed, http.StateHijacked:
                        remove <- conn
                }
@@ -269,7 +272,7 @@ func (srv *Server) Serve(listener net.Listener) error {
        // Manage open connections
        shutdown := make(chan chan struct{})
        kill := make(chan struct{})
-       go srv.manageConnections(add, idle, remove, shutdown, kill)
+       go srv.manageConnections(add, idle, active, remove, shutdown, kill)
 
        interrupt := srv.interruptChan()
        // Set up the interrupt handler
@@ -334,7 +337,7 @@ func DefaultLogger() *log.Logger {
        return log.New(os.Stderr, "[graceful] ", 0)
 }
 
-func (srv *Server) manageConnections(add, idle, remove chan net.Conn, shutdown chan chan struct{}, kill chan struct{}) {
+func (srv *Server) manageConnections(add, idle, active, remove chan net.Conn, shutdown chan chan struct{}, kill chan struct{}) {
        var done chan struct{}
        srv.connections = map[net.Conn]struct{}{}
        srv.idleConnections = map[net.Conn]struct{}{}
@@ -344,6 +347,8 @@ func (srv *Server) manageConnections(add, idle, remove chan net.Conn, shutdown c
                        srv.connections[conn] = struct{}{}
                case conn := <-idle:
                        srv.idleConnections[conn] = struct{}{}
+               case conn := <-active:
+                       delete(srv.idleConnections, conn)
                case conn := <-remove:
                        delete(srv.connections, conn)
                        delete(srv.idleConnections, conn)

```

**Comment by tylerb** [comment]
Ah! The new idleConnetions functionality! Nice find thank you!

**Comment by tylerb** [comment]
Fixed in 9444dcac5b420f65fd5c880051dcb15f87459b81

Thanks @dankinder and @mb0!

**Comment by dankinder** [comment]
Nice, thanks guys!
